### ISSUE TITLE
[java] UnusedPrivateMethod yields false positive for counter-variant arguments

### ISSUE THREAD

**Issue by lgoldstein** [pmd/pmd#770]
Please, prefix the report title with the language it applies to within brackets, such as *[java]* or *[apex]*. If not specific to a language, you can use *[core]*

**Rule Set:**

`rulesets/java/unusedcode.xml/UnusedPrivateMethod`

**Description:**

(PMD 5.6.1) If the invoked private method contains a "weaker" formal parameter type than actual invocation then PMD does not detect its invocation

**Code Sample demonstrating the issue:**

```java

    @Test
    public void testFlipCaseDigit() {
        testNoChangeOnFlip("0123456789");
    }

    @Test
    public void testFlipCaseWhitespace() {
        testNoChangeOnFlip(" \r\t\n");
    }

    @Test
    public void testFlipCaseOtherChars() {
        testNoChangeOnFlip("!@#$%^");
    }

    private static void testNoChangeOnFlip(CharSequence chars) {
        for (int index = 0; index < chars.length(); index++) {
            char expected = chars.charAt(index);
            char actual = CharacterUtils.flipCase(expected);
            assertEquals("Unexpected flipped value", expected, actual);
        }
    }
```

yields `CharacterUtilsTest:44 Rule:UnusedPrivateMethod Priority:3 Avoid unused private methods such as 'testNoChangeOnFlip(CharSequence)'..`

A similar problem occurs with this (more complex) code:

```java
public class ExceptionUtils {
    private static final class ExceptionThrower {
        private static synchronized void spit(Throwable t) {
             ...
        }
   }

    @SuppressWarnings("synthetic-access")
    public static void rethrowException(Throwable t) {
        ExceptionThrower.spit(t);
    }
}
```
which yields `ExceptionThrower:405 Rule:UnusedPrivateMethod Priority:3 Avoid unused private methods such as 'spit(Throwable)'..` - although in this case, the types are the same

**Running PMD through:** *Maven*

**Comment by lgoldstein** [comment]
The same problem seems to occur with generic type:
```java
@Test
    public void testOnlyIdentifierIsUsed() {
        DataSourceEntity<Long> ds = mockEntity(3777347L);   // <<=== the complaint
        Object expected = DataSourceEntityAwareKeyGenerator.DEFAULT.generate(null, null, ds.getDsId());
        Object actual = DataSourceEntityAwareKeyGenerator.DEFAULT.generate(null, null, ds);
        assertSame(expected, actual);   // based on the way SimpleKeyGenerator works when there is only one argument...
    }

    private static <ID extends Serializable & Comparable<ID>> DataSourceEntity<ID> mockEntity(ID id) {
      ...
    }
```

**Comment by jsotuyod** [comment]
@lgoldstein good catch! thanks for the feedback. This was impossible with previous PMD versions, but we may be able to detect these cases properly given the improved type resolution in upcoming releases.

**Comment by adangel** [comment]
This has been fixed with PMD 7.0.0-rc1.
